niceman/utils.py

```diff
@@ -895,4 +896,21 @@ def unicode(s):
     else:
         return __builtin__.unicode(s)
 
+
+def generate_unique_name(pattern, nameset):
+    """Create a unique numbered name from a pattern and a set"""
+    i = 0
+    while True:
+        n = pattern % i
+        i += 1
+        if n not in nameset:
+            return n
+
+
+# http://stackoverflow.com/questions/1151658/python-hashable-dicts
+class HashableDict(dict):
+    """Dict that can be used as keys"""
+    def __hash__(self):
+        return hash(frozenset(self.values()))
```

Comment:
I don't think this would work correctly if there are the same values but different keys... e.g. `{'a': 1}` will have the same hash as `{'b':1}`:
```ipython
In [3]: l({'a': 1})
Out[3]: -8166288242918962057

In [4]: l({'a': 1})
Out[4]: -8166288242918962057

In [5]: l({'b': 1})
Out[5]: -8166288242918962057
```
so should use `.items`:
```ipython
In [6]: l = lambda x: hash(frozenset(x.items()))

In [7]: l({'b': 1})
Out[7]: 966245629147772557

In [8]: l({'a': 1})
Out[8]: -2550060783245333914
```
but if we are getting there... I wonder if we should really make it immutable, at least on its level, as soon as has of it is requested... i.e. so we could compose a dictionary, why not to 
set some `__frozen=True` upon `__hash__` call and then make all `__del__` and `__setitem__` throw an exception if called?  otherwise I am not sure what behaviour would be if it gets changed in place thus invalidating previous assesments etc

I do understand that for current use probably it is ok, but since created/placed to be "generic" -- it better be reusable